test/extended/router/external_certificate.go

```diff
@@ -0,0 +1,769 @@
+		)
+
+		g.BeforeEach(func() {
+			const numRoutes = 5
+			var routeNames []string
+
+			for i := 0; i < numRoutes; i++ {
+				hosts = append(hosts, fmt.Sprintf("host-%d-%s.%s", i, oc.Namespace(), defaultDomain))
+				routeNames = append(routeNames, fmt.Sprintf("route-%d", i))
+			}
+
+			g.By("Creating a TLS certificate secret")
+			secret, rootDerBytes, err = generateTLSCertSecret(oc.Namespace(), "my-tls-secret", corev1.SecretTypeTLS, hosts...)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			_, err = oc.KubeClient().CoreV1().Secrets(oc.Namespace()).Create(context.Background(), secret, metav1.CreateOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Providing router service account permissions to get,list,watch the secret")
+			_, err = oc.KubeClient().RbacV1().Roles(oc.Namespace()).Create(context.Background(),
+				generateSecretReaderRole(oc.Namespace(), "my-tls-secret"), metav1.CreateOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+			_, err = oc.KubeClient().RbacV1().RoleBindings(oc.Namespace()).Create(context.Background(),
+				generateRouterRoleBinding(oc.Namespace()), metav1.CreateOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Creating multiple routes referencing same external certificate")
+			for i := 0; i < numRoutes; i++ {
+				route := generateRouteWithExternalCertificate(oc.Namespace(), routeNames[i], secret.Name, helloPodSvc, hosts[i], routev1.TLSTerminationEdge)
+				_, err = oc.RouteClient().RouteV1().Routes(oc.Namespace()).Create(context.Background(), route, metav1.CreateOptions{})
+				o.Expect(err).NotTo(o.HaveOccurred())
+				routes = append(routes, route)
+			}
+		})
+
+		g.Describe("the router should support external certificate", func() {
+			g.It("and routes are reachable", func() {
+				g.By("Sending https request")
+				for _, route := range routes {
+					hostName, err := getHostnameForRoute(oc, route.Name)
+					o.Expect(err).NotTo(o.HaveOccurred())
+					resp, err := httpsGetCall(hostName, rootDerBytes)
+					o.Expect(err).NotTo(o.HaveOccurred())
+					o.Expect(resp).Should(o.ContainSubstring(helloOpenShiftResponse))
+				}
+			})
+
+			g.Context("and the secret is deleted", func() {
+				g.BeforeEach(func() {
+					g.By("Deleting the secret")
+					err = oc.KubeClient().CoreV1().Secrets(oc.Namespace()).Delete(context.Background(), secret.Name, metav1.DeleteOptions{})
+					o.Expect(err).NotTo(o.HaveOccurred())
+				})
+
+				g.It("then routes are not reachable", func() {
+					g.By("Checking the route status")
+					for _, route := range routes {
+						checkRouteStatus(oc, route.Name, corev1.ConditionFalse, "ExternalCertificateValidationFailed")
+					}
+				})
+
+				g.Context("and re-created again", func() {
+					g.BeforeEach(func() {
+						g.By("Re-creating the deleted secret")
+						_, err = oc.KubeClient().CoreV1().Secrets(oc.Namespace()).Create(context.Background(), secret, metav1.CreateOptions{})
+						o.Expect(err).NotTo(o.HaveOccurred())
+					})
```

Comment:
Does it make sense to make this `BeforeEach`? It will be run only before `then routes are reachable`, won't it?

Reply:
True, but I don't want to _clutter_  `It` blocks, and hence moved all the configuration bits into `BeforeEach` blocks. It would help keep the tests more structured and future additions will be comparatively easier. 

We discussed a bit on those lines here: https://github.com/openshift/origin/pull/29499#discussion_r1990915749

Reply:
I can understand the point of view. However, I don't see how combining these two would clutter the `It` block. The combination of "re-creation of a secret" "check of the route which uses it" seems to be a complete, self contained test case (== `It` block). You did follow this logic in the `Context` below, for instance `should not support external certificate without proper permissions`  1) generates a secret, 2) creates it and 3) verifies the route creation fails, all in one `It` block.
Also, since Arkadeep found this approach questionable as well, chances are that the future readers will have the same question.

Reply:
I've merged the leaf `BeforeEach` and `It` blocks together. Please let me know if it's fine now. Thanks!